tests/sentry/api/serializers/test_alert_rule.py

```diff
@@ -87,3 +94,203 @@ def test_triggers(self):
+        self.setup_project_and_rules()
+        for _ in xrange(150):
+            self.create_alert_rule(projects=self.projects)
+
+        # Test limit above result cap (which is 100), no cursor.
+        with self.feature("organizations:incidents"):
+            request_data = {"cursor": "0:0:0", "limit": "325"}
+            response = self.client.get(
+                path=self.combined_rules_url, data=request_data, content_type="application/json"
+            )
+        assert response.status_code == 200
+        result = json.loads(response.content)
+        assert len(result) == 100
+
+    def test_limit_higher_than_results_no_cursor(self):
+        self.setup_project_and_rules()
+        # Test limit above result count (which is 4), no cursor.
+        with self.feature("organizations:incidents"):
+            request_data = {"cursor": "0:0:0", "limit": "5"}
+            response = self.client.get(
+                path=self.combined_rules_url, data=request_data, content_type="application/json"
+            )
+        assert response.status_code == 200
+        result = json.loads(response.content)
+        assert len(result) == 4
+        self.assert_alert_rule_serialized(self.yet_another_alert_rule, result[0], skip_dates=True)
+        assert result[1]["id"] == six.text_type(self.issue_rule.id)
+        assert result[1]["type"] == "rule"
+        self.assert_alert_rule_serialized(self.other_alert_rule, result[2], skip_dates=True)
+        self.assert_alert_rule_serialized(self.alert_rule, result[3], skip_dates=True)
+
+    def test_limit_as_1_with_paging(self):
+        self.setup_project_and_rules()
+
+        # Test Limit as 1, no cursor:
+        with self.feature("organizations:incidents"):
+            request_data = {"cursor": "0:0:0", "limit": "1"}
+            response = self.client.get(
+                path=self.combined_rules_url, data=request_data, content_type="application/json"
+            )
+        assert response.status_code == 200
+
+        result = json.loads(response.content)
+        assert len(result) == 1
+        self.assert_alert_rule_serialized(self.yet_another_alert_rule, result[0], skip_dates=True)
+
+        links = requests.utils.parse_header_links(
+            response.get("link").rstrip(">").replace(">,<", ",<")
+        )
+        next_cursor = links[1]["cursor"]
+
+        # Test Limit 1, next page of previous request:
+        with self.feature("organizations:incidents"):
+            request_data = {"cursor": next_cursor, "limit": "1"}
+            response = self.client.get(
+                path=self.combined_rules_url, data=request_data, content_type="application/json"
+            )
+        assert response.status_code == 200
+
+        result = json.loads(response.content)
+        assert len(result) == 1
+        assert result[0]["id"] == six.text_type(self.issue_rule.id)
+        assert result[0]["type"] == "rule"
+
+    def test_limit_as_2_with_paging(self):
```

Comment:
For the 3 calls here the cursors end up looking like:
```
0:0:0 1970-01-01 00:00:00+00:00
1576629180:2:0 2019-12-18 00:33:00+00:00
1576629180:4:0 2019-12-18 00:33:00+00:00
```

The second two are all for the same time, and are just differentiated based on offset. This is caused by the alert rules being all created too close together.

If you create the rules a minute apart each, Calls end up looking like

```
0:0:0 1970-01-01 00:00:00+00:00
1576629205:2:0 2019-12-18 00:33:25+00:00
```

Then we fail with 
```
________________________________________ CombinedRuleSerializerTest.test_limit_as_2_with_paging ________________________________________
Traceback (most recent call last):
  File "/Users/dfuller/code/sentry/tests/sentry/api/serializers/test_alert_rule.py", line 289, in test_limit_as_2_with_paging
    assert len(result) == 2
AssertionError: assert 1 == 2
 +  where 1 = len([{'aggregation': 0, 'aggregations': [0], 'alertThreshold': 0, 'dataset': 'events', ...}])
```


You can reproduce this with this patch
```
diff --git a/tests/sentry/api/serializers/test_alert_rule.py b/tests/sentry/api/serializers/test_alert_rule.py
index 1f51a5f978..a4cbe71237 100644
--- a/tests/sentry/api/serializers/test_alert_rule.py
+++ b/tests/sentry/api/serializers/test_alert_rule.py
@@ -16,6 +16,7 @@ from sentry.incidents.logic import create_alert_rule, create_alert_rule_trigger
 from sentry.incidents.models import AlertRuleThresholdType
 from sentry.snuba.models import QueryAggregations
 from sentry.testutils import TestCase, APITestCase
+from sentry.testutils.helpers.datetime import before_now


 class BaseAlertRuleSerializerTest(object):
@@ -103,8 +104,12 @@ class CombinedRuleSerializerTest(BaseAlertRuleSerializerTest, APITestCase, TestC
         self.project = self.create_project(organization=self.org, teams=[self.team], name="Bengal")
         self.login_as(self.user)
         self.projects = [self.project, self.create_project()]
-        self.alert_rule = self.create_alert_rule(projects=self.projects)
-        self.other_alert_rule = self.create_alert_rule(projects=self.projects)
+        self.alert_rule = self.create_alert_rule(
+            projects=self.projects, date_added=before_now(minutes=4)
+        )
+        self.other_alert_rule = self.create_alert_rule(
+            projects=self.projects, date_added=before_now(minutes=3)
+        )
         self.issue_rule = self.create_issue_alert_rule(
             data={
                 "project": self.project,
@@ -112,14 +117,17 @@ class CombinedRuleSerializerTest(BaseAlertRuleSerializerTest, APITestCase, TestC
                 "conditions": [],
                 "actions": [],
                 "actionMatch": "all",
-            }
+            },
+            date_added=before_now(minutes=2),
+        )
+        self.yet_another_alert_rule = self.create_alert_rule(
+            projects=self.projects, date_added=before_now(minutes=1)
         )
-        self.yet_another_alert_rule = self.create_alert_rule(projects=self.projects)
         self.combined_rules_url = "/api/0/projects/{0}/{1}/combined-rules/".format(
             self.org.slug, self.project.slug
         )

-    def create_issue_alert_rule(self, data):
+    def create_issue_alert_rule(self, data, date_added=None):
         """data format
         {
             "project": project
@@ -145,6 +153,8 @@ class CombinedRuleSerializerTest(BaseAlertRuleSerializerTest, APITestCase, TestC
             rule.data["conditions"] = data["conditions"]
         if data.get("frequency"):
             rule.data["frequency"] = data["frequency"]
+        if date_added:
+            rule.date_added = date_added
         rule.save()
         return rule

diff --git a/src/sentry/testutils/factories.py b/src/sentry/testutils/factories.py
index 3e2aafe594..80d46ec834 100644
--- a/src/sentry/testutils/factories.py
+++ b/src/sentry/testutils/factories.py
@@ -939,11 +939,12 @@ class Factories(object):
         threshold_period=1,
         include_all_projects=False,
         excluded_projects=None,
+        date_added=None,
     ):
         if not name:
             name = petname.Generate(2, " ", letters=10).title()

-        return create_alert_rule(
+        alert_rule = create_alert_rule(
             organization,
             projects,
             name,
@@ -954,6 +955,9 @@ class Factories(object):
             include_all_projects=include_all_projects,
             excluded_projects=excluded_projects,
         )
+        if date_added is not None:
+            alert_rule.update(date_added=date_added)
+        return alert_rule
```

I'm pretty sure this is caused by using `gte` rather than `lte`

Reply:
Wow Dan, good catch and thorough review. I really appreciate it.

This uncovered a few issues - the major being that we were not using microsecond data in the cursor. The microseconds were formatted as a decimal, and the build cursor function was casting it as in int. The other being that gte was being used - and only working because all the alert rules were clumped into the same second. When switched to lte, the lack of microsecond precision made paging fail because microseconds would make the rule's date_added greater than the cursor's date value.

We now multiply our cursor values by 100000 to keep the microseconds before the cursor (and leaving the int casting in the cursor alone). The cursor's in that particular test now end up being something like:
`
('1576697690152634:0:0', datetime.datetime(2019, 12, 18, 19, 34, 50, 152634, tzinfo=<UTC>))
('next cursor:', '1576697389587382:0:0')
(u'1576697389587382:0:0', datetime.datetime(2019, 12, 18, 19, 29, 49, 587382, tzinfo=<UTC>))
('next cursor:', '1576697329563132:1:0')
(u'1576697329563132:1:0', datetime.datetime(2019, 12, 18, 19, 28, 49, 563132, tzinfo=<UTC>))
`

Which is now correct afaict.  This also helped me get to the bottom of why the offset test was acting so wonky (why was my offset 2?!). It really cleared everything up. Thanks!